som-interpreter-bc/src/interpreter.rs

```diff
@@ -195,12 +196,76 @@ impl Interpreter {
                     };
                     let signature = universe.lookup_symbol(symbol);
                     let nb_params = nb_params(signature);
-                    let method = self
-                        .stack
-                        .iter()
-                        .nth_back(nb_params)
-                        .unwrap()
-                        .lookup_method(universe, symbol);
+                    let method = {
+                        let receiver = self.stack.iter().nth_back(nb_params)?;
+                        let receiver_class = receiver.class(universe);
+                        match frame.borrow().kind() {
+                            FrameKind::Block { block } => {
+                                let mut inline_cache_receiver =
+                                    block.inline_cache_receiver.borrow_mut();
+                                let mut inline_cache_invocable =
+                                    block.inline_cache_invocable.borrow_mut();
+                                // SAFETY: this access is actually safe because the bytecode compiler
+                                // makes sure the cache has as many entries as there are bytecode instructions,
+                                // therefore we can avoid doing any redundant bounds checks here.
+                                let maybe_found_receiver = unsafe {
+                                    inline_cache_receiver.get_unchecked_mut(bytecode_idx)
+                                };
+                                let maybe_found_invocable = unsafe {
+                                    inline_cache_invocable.get_unchecked_mut(bytecode_idx)
+                                };
+
+                                match (maybe_found_receiver, maybe_found_invocable) {
+                                    (receiver, Some(method))
+                                        if *receiver == receiver_class.as_ptr() =>
+                                    {
+                                        Some(Rc::clone(method))
+                                    }
+                                    (receiver, method) => {
+                                        let found = receiver_class.borrow().lookup_method(symbol);
+                                        *receiver = receiver_class.as_ptr();
+                                        *method = found.clone();
+                                        found
+                                    }
+                                }
+                            }
+                            FrameKind::Method { method, .. } => {
```

Comment:
This is something I've noticed in general when working on som-rs, there's a lot of duplication between Methods and Blocks, but they're almost the same thing right? Could they be unified somehow? (this is outside the scope of those inline caching changes, but seeing it here reminded me again)

Reply:
They are unified in all other SOMs: a block refers to a method. And then everything follows from there.

Reply:
Yeah, it is true that there are plenty of places where the code could be a lot cleaner than how it currently is.  
I think this is because I originally wrote this code when the interpreter wasn't working yet and I just wanted to get it working before doing refactoring, but I ended up never doing them.  
These two branches can definitely be unified, the only difference between them is where the location of the inline cache storage.